lib/src/lock_file.dart

```diff
@@ -308,25 +314,80 @@ Generated by pub on ${DateTime.now()}.''';
     return '${JsonEncoder.withIndent('  ').convert(packageConfig.toJson())}\n';
   }
 
+  /// Throws a [DataException] if the content-hash of some hosted package locked
+  /// in this lock file differs from the one in the [cache].
+  void checkContentHashes(SystemCache cache) {
+    for (final id in packages.values) {
+      var description = id.description;
+      if (description is ResolvedHostedDescription) {
+        Uint8List? cachedHash =
+            description.description.source.sha256FromCache(id, cache);
+        if (cachedHash == null) {
+          // This can happen if we resolve from a server not providing hashes,
+          // but we are not downloading the archives. (eg. for a
+          // dependency_services run).
+          log.fine(
+              'No hash of downloaded archive for ${id.name}-${id.version} found');
+        } else if (description.sha256 == null) {
+          // We have resolved from a server without archive_sha256 in the
+          // version listing. Use the sha256 from the archive instead.
+          description = description.withSha256(cachedHash);
+        } else {
+          // TODO(Should we really not fail here???).
```

Comment:
Let's make a test case or two that demonstrates this.

I'm thinking that this is possibly a reason to:
 * Change the lockfile
 * Re-download the file

Or maybe it's an internal error, or maybe I missed something obvious.